File: rubix/flutter/packages/rubix_sdui/lib/src/widgets/components/accent.dart
/// Accent resolution for KPI tiles and chart series.
///
/// Dart port of `packages/starter-ui-sdui-react/src/renderer/accent.ts`.
/// The hash function and intent map must stay identical so the same
/// IR renders with the same accent on web and Flutter.
library;

import '../sdui_theme.dart';

const _intentMap = <String, SduiAccent>{
  'primary':  SduiAccent.leaf,
  'positive': SduiAccent.leaf,
  'good':     SduiAccent.leaf,
  'info':     SduiAccent.sky,
  'warn':     SduiAccent.warn,
  'warning':  SduiAccent.warn,
  'energy':   SduiAccent.sun,
  'cool':     SduiAccent.aqua,
};

const _autoRotation = <SduiAccent>[
  SduiAccent.leaf,
  SduiAccent.aqua,
  SduiAccent.sun,
  SduiAccent.sky,
];

int _hash(String s) {
  var h = 0;
  for (var i = 0; i < s.length; i++) {
    // Match JS `(h * 31 + c) | 0` — clamp to 32-bit signed.
    h = (h * 31 + s.codeUnitAt(i)).toSigned(32);
  }
  return h.abs();
}

/// Returns the [SduiAccent] for a component node.
///
/// Resolution order: explicit `accent` → mapped `intent` → hash of
/// `id` (skips `warn`, reserved for explicit intent).
SduiAccent resolveAccent(Map<String, Object?> node) {
  final accent = node['accent'];
  if (accent is String) {
    for (final v in SduiAccent.values) {
      if (v.name == accent) return v;
    }
  }
  final intent = node['intent'];
  if (intent is String && _intentMap.containsKey(intent)) {
    return _intentMap[intent]!;
  }
  final id = node['id'];
  final seed = id is String ? id : '';
  return _autoRotation[_hash(seed) % _autoRotation.length];
}

/// Rotation by index — used by `kpi_grid` so siblings read
/// leaf → aqua → sun → sky in order.
SduiAccent accentByIndex(int i) => _autoRotation[i % _autoRotation.length];

File: packages/starter-ui-sdui-react/src/renderer/accent.ts
// Accent rotation for KPI tiles and chart series. Maps a logical
// accent name to the rubix theme tokens (var(--color-*)) so SDUI
// visuals follow the active palette (nube / ocean / sunset) without
// renderer changes.
//
// Server-side IR can pin a tile to a specific accent via `node.accent`
// or `intent`; otherwise we derive one from `node.id` so siblings get
// stable, visually-distinct colors across re-renders.

export const SDUI_ACCENTS = ["leaf", "aqua", "sun", "sky", "warn"] as const;
export type SduiAccent = (typeof SDUI_ACCENTS)[number];

const ACCENT_VAR: Record<SduiAccent, string> = {
  leaf: "var(--color-leaf)",
  aqua: "var(--color-aqua)",
  sun: "var(--color-sun)",
  sky: "var(--color-sky)",
  warn: "var(--color-warn)",
};

const INTENT_MAP: Record<string, SduiAccent> = {
  primary: "leaf",
  positive: "leaf",
  good: "leaf",
  info: "sky",
  warn: "warn",
  warning: "warn",
  energy: "sun",
  cool: "aqua",
};

function hash(input: string): number {
  let h = 0;
  for (let i = 0; i < input.length; i++) h = (h * 31 + input.charCodeAt(i)) | 0;
  return Math.abs(h);
}

export function resolveAccent(node: {
  id?: unknown;
  accent?: unknown;
  intent?: unknown;
}): SduiAccent {
  if (typeof node.accent === "string" && (SDUI_ACCENTS as readonly string[]).includes(node.accent)) {
    return node.accent as SduiAccent;
  }
  if (typeof node.intent === "string" && INTENT_MAP[node.intent]) {
    return INTENT_MAP[node.intent]!;
  }
  const seed = typeof node.id === "string" ? node.id : "";
  // Skip `warn` in the auto-rotation — reserve it for explicit intent.
  const palette = SDUI_ACCENTS.slice(0, 4);
  return palette[hash(seed) % palette.length]!;
}

export function accentVar(accent: SduiAccent): string {
  return ACCENT_VAR[accent];
}

export function accentByIndex(index: number): SduiAccent {
  const palette = SDUI_ACCENTS.slice(0, 4);
  return palette[index % palette.length]!;
}
